cognite/client/_api/datapoints.py

```diff
@@ -517,6 +498,248 @@ def __init__(self, config: ClientConfig, api_version: str | None, cognite_client
+                raise ValueError(
+                    "When iterating datapoints, the options 'include_outside_points' and 'limit' are not supported."
+                )
+
+        for ident_type, queries in qs_per_id_type.items():
+            if dupes := find_duplicates(q.identifier.as_primitive() for q in queries):
+                raise ValueError(
+                    f"When iterating datapoints, identifiers must be unique! Duplicates found for {ident_type}: {dupes}"
+                )
+        original_order = [q.identifier for q in user_queries]
+
+        parsed_queries = _FullDatapointsQuery(
+            start=start,
+            end=end,
+            id=qs_per_id_type["id"],
+            external_id=qs_per_id_type["external_id"],
+            instance_id=qs_per_id_type["instance_id"],
+            aggregates=aggregates,
+            granularity=granularity,
+            timezone=timezone,
+            target_unit=target_unit,
+            target_unit_system=target_unit_system,
+            limit=request_limit,
+            include_outside_points=False,
+            ignore_unknown_ids=ignore_unknown_ids,
+            include_status=include_status,
+            ignore_bad_datapoints=ignore_bad_datapoints,
+            treat_uncertain_as_bad=treat_uncertain_as_bad,
+        ).parse_into_queries()
+        self.query_validator(parsed_queries)
+        alive_queries = {q.identifier: q for q in parsed_queries}
+        alive_queries = {ident: alive_queries[ident] for ident in original_order}  # ..and sort
+
+        while alive_queries:
+            to_fetch_queries = list(itertools.islice(alive_queries.values(), chunk_size_time_series))
+            fetcher = self._select_dps_fetch_strategy(to_fetch_queries)(
+                self, to_fetch_queries, self._config.max_workers
+            )
+            dps_lst: DatapointsArrayList | DatapointsList = (
+                fetcher.fetch_all_datapoints_numpy() if return_arrays else fetcher.fetch_all_datapoints()
+            )
+            for query in to_fetch_queries:
+                ident = query.identifier
+                dps = dps_lst.get(**ident.as_dict(camel_case=False))
+                if isinstance(dps, list):
+                    raise RuntimeError(
+                        "When iterating datapoints, identifiers must be unique! You can not get around this by passing "
+                        "several of [id, external_id, instance_id] for the same underlying time series."
+                    )
+                # Update query.start for next iteration if ts is not yet exhausted:
+                if dps and len(dps) == request_limit:
+                    new_start = cast(int, dps[-1].timestamp) + 1
+                    if query.end_ms > new_start:
+                        query.start = new_start  # manual cursoring ftw
+                        continue
+                alive_queries.pop(ident)
```

Comment:
Just for readability, could this be pulled out to a helper. Makes the overall loop easier to read (at least for me :P) 
```suggestion
self._update_query(alive_queries, dps_lst, request_limit, to_fetch_queries)
```

Reply:
Sure 👍 Like a manual cursoring method 😅